src/utils/openLink.js

```diff
@@ -5,19 +5,17 @@ import * as WebBrowser from 'expo-web-browser';
 import type { BrowserPreference, GlobalSettingsState } from '../types';
 
 /** Open a URL in the in-app browser. */
-// TODO(#4146): Take a URL object, not a string.
-export function openLinkEmbedded(url: string): void {
+export function openLinkEmbedded(url: URL): void {
   if (Platform.OS === 'ios') {
-    WebBrowser.openBrowserAsync(encodeURI(url));
+    WebBrowser.openBrowserAsync(encodeURI(url.toString()));
```

Comment:
Hmm, does this version indeed work for fixing #4136?

Based on the discussion in that thread, I'd expect the issue to be present as long as this `encodeURI` call is here.

Once the URL always comes from a `URL` object, we should be able to just remove the `encodeURI` call; that should fix #4136, and when the URL is coming from a `URL` object it shouldn't reintroduce #3315.

I guess this illustrates that before closing #4136 we should do some actual testing of the fix :wink: Per the issue description:

> A key step in fixing this is going to be just end-to-end testing: make a URL filled with a ton of these characters, post it in a Zulip message, try following that link, and see what URL actually comes through.

Reply:
Ah right, thanks for the catch! I'll be sure to test my next revision manually. 🙂

Do you think we have to round-trip through `new URL` and remove the `encodeURI` in the same commits, or is it OK to have a commit that does `new URL` and still has `encodeURI` in place?

Reply:
> or is it OK to have a commit that does `new URL` and still has `encodeURI` in place?

I think that'd be OK, as long as we're merging at the same time a subsequent commit that removes the `encodeURI`.

It would make #4136 apply more widely, so I wouldn't want to release in that state and hence wouldn't want to merge a branch that left things in that state. But the `new URL` change *ought* to be harmless, and the fact that it isn't is because of an existing bug (the use of `encodeURI`), so I'm OK with having a commit in the history that does that first.